util/launcher/src/migrations/set_2019_block_cycle_zero.rs

```diff
@@ -0,0 +1,95 @@
+use ckb_app_config::StoreConfig;
+use ckb_db_migration::{Migration, ProgressBar, ProgressStyle};
+use ckb_db_schema::COLUMN_EPOCH;
+use ckb_store::{ChainDB, ChainStore};
+use ckb_types::{
+    core::hardfork::HardForks,
+    packed,
+    prelude::{Entity, FromSliceShouldBeOk, Pack, Reader},
+};
+
+const VERSION: &str = "20231101000000";
+pub struct BlockExt2019ToZero {
+    hardforks: HardForks,
+}
+
+impl BlockExt2019ToZero {
+    pub fn new(hardforks: HardForks) -> Self {
+        BlockExt2019ToZero { hardforks }
+    }
+}
+
+impl Migration for BlockExt2019ToZero {
+    fn run_in_background(&self) -> bool {
+        true
+    }
+
+    fn migrate(
+        &self,
+        db: ckb_db::RocksDB,
+        pb: std::sync::Arc<dyn Fn(u64) -> ProgressBar + Send + Sync>,
+    ) -> Result<ckb_db::RocksDB, ckb_error::Error> {
+        let chain_db = ChainDB::new(db, StoreConfig::default());
+        let limit_epoch = self.hardforks.ckb2021.rfc_0032();
+
+        eprintln!("begin to run block_ext 2019 to zero migrate...");
+        if limit_epoch == 0 {
+            return Ok(chain_db.into_inner());
+        }
+
+        let epoch_number: packed::Uint64 = limit_epoch.pack();
+        if let Some(epoch_hash) = chain_db.get(COLUMN_EPOCH, epoch_number.as_slice()) {
```

Comment:
There are some special cases not covered here：

If the current node synchronization has not exceeded hardfork epoch, this judgment is problematic.

So getting the first header here still requires some operations.
1. Get the current tip header
2. Compare the epoch of the tip header with the epoch of the hardfork switch, and then determine whether the first header is the tip header or the header corresponding to the previous hash of the hardfork switch.